src/Servers/Kestrel/Core/src/Internal/Http3/Http3HeadersEnumerator.cs

```diff
@@ -0,0 +1,152 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+using Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http3
+{
+    internal sealed class Http3HeadersEnumerator : IEnumerator<KeyValuePair<string, string>>
+    {
+        private enum HeadersType : byte
+        {
+            Headers,
+            Trailers,
+            Untyped
+        }
+        private HeadersType _headersType;
+        private HttpResponseHeaders.Enumerator _headersEnumerator;
+        private HttpResponseTrailers.Enumerator _trailersEnumerator;
+        private IEnumerator<KeyValuePair<string, StringValues>>? _genericEnumerator;
+        private StringValues.Enumerator _stringValuesEnumerator;
+        private bool _hasMultipleValues;
+        private KnownHeaderType _knownHeaderType;
+
+        public Func<string, Encoding?> EncodingSelector { get; set; } = KestrelServerOptions.DefaultHeaderEncodingSelector;
+
+        public int QPackStaticTableId => GetResponseHeaderStaticTableId(_knownHeaderType);
```

Comment:
Where/how is this used?

Reply:
The pattern is copied from Http2HeadersEnumerator and will be used when implementing https://github.com/dotnet/aspnetcore/issues/33980.